src/util/db/dbid.h

```diff
@@ -1,126 +1,166 @@
+            : DbId(NoAssert{}, valueOf(std::move(variant))) {
     }
 
-    bool isValid() const {
+    /// Explicit checked conversion from inner value
+    ///
+    /// Creates an instance from the provided inner value. The value
+    /// might be either valid or invalid. Triggers a debug assertion
+    /// if an invalid value doesn't equal the predefined constant.
+    explicit DbId(value_type value)
+            : DbId(NoAssert{}, value) {
+        DEBUG_ASSERT(isValid() || m_value == kInvalidValue);
+    }
+
+    /// Explicit unchecked conversion from inner value
+    ///
+    /// Creates an instance with the provided inner value without
+    /// any consistency checks that could trigger an assertion.
+    static constexpr DbId fromValueUnchecked(value_type value) {
+        return DbId(NoAssert{}, value);
+    }
+
+    constexpr bool isValid() const {
         return isValidValue(m_value);
     }
 
+    /// Explicit unchecked conversion into the inner value
+    ///
+    /// Returns the inner value, either valid or invalid.
+    constexpr value_type valueMaybeInvalid() const {
+        return m_value;
+    }
+
+    /// Explicit checked conversion into the inner value
+    ///
+    /// Returns the inner value. Triggers a debug assertion when
+    /// invoked on invalid instances.
     value_type value() const {
+        DEBUG_ASSERT(isValid());
         return m_value;
     }
 
+    /// Implicit checked conversion of the inner value
+    ///
+    /// Returns the inner value. Triggers a debug assertion when
+    /// invoked on invalid instances.
+    operator value_type() const {
+        return value();
+    }
+
     std::size_t hash() const {
         return std::hash<value_type>()(m_value);
     }
 
-    typedef std::function<std::size_t (const DbId& dbid)> hash_fun_t;
+    typedef std::function<std::size_t(const DbId& dbid)> hash_fun_t;
     static std::size_t hash_fun(const DbId& dbid) {
         return dbid.hash();
     }
 
-    // This function should be used for value binding in DB queries
-    // with bindValue().
+    /// Convert into QVariant
+    ///
+    /// This function is used for value binding in DB queries with bindValue().
     QVariant toVariant() const {
-        return QVariant(m_value);
+        return isValid() ? QVariant{m_value} : QVariant{};
     }
 
+    /// Convert into QString
+    ///
+    /// This function is used for formatting comma-separated lists of ids in DB queries.
     QString toString() const {
-        return QString::number(m_value);
-    }
-
-    friend bool operator==(const DbId& lhs, const DbId& rhs) {
-        return lhs.m_value == rhs.m_value;
+        return isValid() ? QString::number(m_value) : QString{};
     }
 
-    friend bool operator!=(const DbId& lhs, const DbId& rhs) {
-        return lhs.m_value != rhs.m_value;
+  private:
+    /// Type tag for overloaded constructor that doesn't assert
+    struct NoAssert {};
+    constexpr DbId(NoAssert, value_type value)
```

Comment:
What is the use case for this? 
I think in all cases we want the assertion.

Reply:
I think each invocation can be replaced with a plain 
```
: m_value(value)
```
No need for an ambiguous overload and the empty struct workaround.

Reply:
No

Reply:
Why not?

Reply:
Because it conflicts with the public conversion ctor that would have the same signature.